backend/mora/common.py

```diff
@@ -52,26 +53,48 @@ def get(self, objid):
 
 
 class RequestType(enum.Enum):
-    CREATE = 0
-    EDIT = 1
+    CREATE, EDIT, TERMINATE = range(3)
 
 
 # The handler mapping is populated by each individual active RequestHandler
-HANDLERS = {}
+HANDLERS_BY_ROLE_TYPE = {}
+HANDLERS_BY_FUNCTION_KEY = {}
+FUNCTION_KEYS = {}
 
 
-def register_request_handler(name):
-    """Decorator to register a writing request handler"""
-    def wrapper(cls):
-        HANDLERS[name] = cls
+class _RequestHandlerMeta(abc.ABCMeta):
+    '''Metaclass for automatically registering handlers
+
+    '''
+
+    @staticmethod
+    def __new__(mcls, name, bases, namespace):
+        cls = super().__new__(mcls, name, bases, namespace)
+
+        if cls.role_type is not None:
+            # sanity check
+            assert cls.role_type not in HANDLERS_BY_ROLE_TYPE
+
+            HANDLERS_BY_ROLE_TYPE[cls.role_type] = cls
+
+        if cls.function_key is not None:
+            # sanity checks
+            assert cls.function_key not in HANDLERS_BY_FUNCTION_KEY
+            assert cls.role_type is not None
+
+            HANDLERS_BY_FUNCTION_KEY[cls.function_key] = cls
+            FUNCTION_KEYS[cls.role_type] = cls.function_key
+
         return cls
-    return wrapper
 
 
-class RequestHandler(abc.ABC):
+class RequestHandler(metaclass=_RequestHandlerMeta):
 
     __slots__ = 'request', 'request_type', 'payload', 'uuid'
 
+    role_type = None
+    function_key = None
```

Comment:
I still don't think it makes sense to have this here. Function key _only_ makes sense for org funk handlers, so the other handler types shouldn't know or care about it.